### ISSUE TITLE
Consume the new C# 7.3 constraints if/where possible

### ISSUE THREAD

**Issue by tannergooding** [dotnet/corefx#28612]
C# 7.3 is adding support for three new constraints:
* `unmanaged`
* `Delegate`/`MulticastDelegate`
* `Enum`

There are several APIs in CoreFX/CoreCLR which are emulating these constraints via runtime checks.

If possible (since it would be a compile time breaking change), it would be nice to update these methods to use the proper constraints where appropriate.

A simple `git grep TDelegate` and `git grep TEnum` will display the majority of these APIs.

**Comment by tannergooding** [comment]
Some examples of this are:
* `System.Linq.Expressions.Expression<TDelegate>`
* `System.Runtime.InteropServices.Marshal.GetFunctionPointerForDelegate<TDelegate>`
* `System.Security.Cryptography.Asn1.AsnReader.GetEnumeratedValue<TEnum>`
* `System.ComponentModel.Composition.Expectations.GetEnumValues<TEnum>`
* `System.Enum.Parse<TEnum>`
* etc

**Comment by joshfree** [comment]
cc: @VSadov @jaredpar

**Comment by jkotas** [comment]
> it would be a compile time breaking change

I do not think we will be able to add these constrains for any existing APIs because of this. I think this issue can be closed.

**Comment by tannergooding** [comment]
@jkotas, to be clear, many of these APIs have runtime checks that validate the constraint, so while you can do something like `GetFunctionPointerForDelegate(int)` it will always throw an exception at runtime.

My thought was: Given that any code not matching the more exact constraint would always fail anyways, it may be worth doing a compat-bar check on a case-by-case basis to see if we can add the more exact constraint.

**Comment by tannergooding** [comment]
That is, any code not matching the constraint is either:
* Doing control flow via exception handling
or
* Has dead code that never executes

**Comment by jkotas** [comment]
I do not think your list is complete. It affects generic wrappers as well:

```
void foo<T>()
{
     ... Expression<T> ...
}
```

This will not compile unless I add the Delegate constrain to my method. And then all callers of my methods have to add the Delegate constrain too. I do not see why we would inflict this pain onto people.
